src/core/lwt.mli

```diff
@@ -21,502 +22,1785 @@
+    which provides some promise-friendly patterns, like streams and mvars. This
+    consists of the modules {!Lwt_list}, {!Lwt_stream}, {!Lwt_result},
+    {!Lwt_mutex}, {!Lwt_condition}, {!Lwt_mvar}, {!Lwt_pool}, and {!Lwt_switch}.
+
+    Above that is the Lwt Unix binding, where I/O begins. This includes the
+    module {!Lwt_main}, including the all-important {!Lwt_main.run}. The rest of
+    the Unix binding consists of functions, each one of which...
+
+    - ...starts a background I/O operation,
+    - creates a promise for it and gives it to you,
+    - registers with {!Lwt_main.run}, so if you attach callbacks to the promise,
+      they will be called when the I/O operation completes.
+
+    The functions are grouped into modules:
+
+    - {!Lwt_unix} for Unix system calls.
+    - {!Lwt_bytes} for Unix system calls on bigarrays.
+    - {!Lwt_io} for [Pervasives]-like high-level channels, TCP servers, etc.
+    - {!Lwt_process} for managing subprocesses.
+    - {!Lwt_preemptive} for spawning system threads.
+    - Miscellaneous modules {!Lwt_daemon}, {!Lwt_gc}, {!Lwt_log}, {!Lwt_engine},
+      {!Lwt_throttle}, {!Lwt_timeout}, {!Lwt_sys}.
+
+    Finally, there are some modules, which belong in the core library or the
+    Unix binding, but are kept separate for historical or administrative
+    reasons. These are {!Lwt_react}, {!Lwt_ssl}, and {!Lwt_glib}. {!Lwt_react}
+    is belongs in the core, and the other two work like the Unix binding.
+
+    Warning! Introductory material ends and detailed reference begins! *)
+
+
+
+(** {2 Fundamentals} *)
+
+(** {3 Promises} *)
 
 type +'a t
-  (** The type of threads returning a result of type ['a]. *)
+(** Promises for values of type ['a].
+
+    A {b promise} is a memory cell that is always in one of three {b states}:
+
+    - {e fulfilled}, and containing one value of type ['a],
+    - {e rejected}, and containing one exception, or
+    - {e pending}, in which case it may become fulfilled or rejected later.
+
+    A {e resolved} promise is one that is either fulfilled or rejected, i.e. not
+    pending. Once a promise is resolved, its content cannot change. So, promises
+    are {e write-once references}. The only possible state changes are (1) from
+    pending to fulfilled and (2) from pending to rejected.
+
+    Promises are typically “read” by attaching {b callbacks} to them. The most
+    basic functions for that are {!Lwt.bind}, which attaches a callback that is
+    called when a promise becomes fulfilled, and {!Lwt.catch}, for rejection.
+
+    Promises variables of this type, ['a Lwt.t], are actually {b read-only} in
+    Lwt. Separate {e resolvers} of type ['a ]{!Lwt.u} are used to write to them.
+    Promises and their resolvers are created together by calling {!Lwt.task}.
+    There is one exception to this: most promises can be {e canceled} by calling
+    {!Lwt.cancel}, without going through a resolver. *)
```

Comment:
I'm not sure why. `Lwt.u` also aren't writeable promises, they are the write ends of each promise for which you happen to have an exposed `Lwt.u`. All (pending) promises are writeable, it's just the `Lwt.t` reference to a promise doesn't allow writing through it. Anyway, this paragraph's idea structure is:

1. `Lwt.t` read-only
2. The write end is `Lwt.u`, which is usually obtained by `task`
3. Cancelation is an exception to (1), and weakly to (2) (to the implication that you *need* the `Lwt.u` to write to a promise).

It would seem like an important omission to leave (3) out of `Lwt.t` by moving it to `Lwt.u`, as it's very important for reasoning about promises that the only way to write to them through an `Lwt.t` is by cancelation.

How does this read to you?